lib/staggered_quark_smearing.cu

```diff
@@ -0,0 +1,221 @@
+#include <dslash.h>
+#include <worker.h>
+#include <dslash_helper.cuh>
+#include <color_spinor_field_order.h>
+#include <gauge_field_order.h>
+#include <color_spinor.h>
+#include <dslash_helper.cuh>
+#include <index_helper.cuh>
+#include <gauge_field.h>
+#include <uint_to_char.h>
+
+#include <dslash_policy.cuh>
+#include <kernels/staggered_quark_smearing.cuh>
+
+/**
+   This is the staggered (two-link) Laplacian derivative.
+*/
+
+namespace quda
+{
+
+  template <typename Arg> class StaggeredQSmear : public Dslash<staggered_qsmear, Arg>
+  {
+    using Dslash = Dslash<staggered_qsmear, Arg>;
+    using Dslash::arg;
+    using Dslash::in;
+
+  public:
+    StaggeredQSmear(Arg &arg, const ColorSpinorField &out, const ColorSpinorField &in) : Dslash(arg, out, in) {}
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      Dslash::setParam(tp);
+
+      // reset threadDimMapLower and threadDimMapUpper when t0 is given
+      // partial replication of dslash::setFusedParam()
+      if( arg.is_t0_kernel )
+      {
+        int prev = -1;
+        for( int i=0; i<4; i++ )
+        {
+          arg.threadDimMapLower[i] = 0;
+          arg.threadDimMapUpper[i] = 0;
+          if( !(arg.commDim[i]) ) continue;
+          arg.threadDimMapLower[i] = ( prev>=0 ? arg.threadDimMapUpper[prev] : 0 );
+          arg.threadDimMapUpper[i] = arg.threadDimMapLower[i] + this->Nface() * (in.GhostFaceCB())[i];
+          prev = i;
+        }
+      }
+
+      // operator is Hermitian so do not instantiate dagger
+      if (arg.nParity == 1) {
+        if(arg.is_t0_kernel) //FIXME we reuse xpay non-type template parameter for t0 option!
```

Comment:
Is this really a `FIXME` or just helpful information?  I think abusing the `xpay` template like this seems like a reasonable thing to do.

Reply:
Actually thinking about this a bit more, I think we might want to fix this, so that it doesn't use the `xpay` template.  

(The `xpay` template is something that I think I might want to explore removing in the future.  For example, making it a run-time option when fast-compilation for the dslash is enabled).

For the present case here, maybe just make the `is_t0_kernel` parameter a run-time variable.  Looking at the code, it doesn't look like this really needs to be a compile-time parameter and run-time should be sufficient.  This will of course also reduce compilation time.

Reply:
fixed